analyzer/block/block.go

```diff
@@ -0,0 +1,349 @@
+			ctx,
+			queries.PickBlocksForProcessing,
+			b.analyzerName,
+			from,
+			lockExpiryMinutes,
+			blocksBatchSize,
+		)
+	default:
+		rows, err = b.target.Query(
+			ctx,
+			queries.PickBlocksForProcessingRange,
+			b.analyzerName,
+			from,
+			to,
+			lockExpiryMinutes,
+			blocksBatchSize,
+		)
+	}
+	switch err {
+	case nil:
+		// Continues below.
+	case pgx.ErrNoRows:
+		// No blocks to process.
+		return heights, nil
+	default:
+		return nil, err
+	}
+
+	defer rows.Close()
+	for rows.Next() {
+		var height uint64
+		if err = rows.Scan(
+			&height,
+		); err != nil {
+			return nil, fmt.Errorf("scanning returned height: %w", err)
+		}
+		heights = append(heights, height)
+	}
+	return heights, nil
+}
+
+// Start starts the block analyzer.
+func (b *blockBasedAnalyzer) Start() {
+	ctx := context.Background()
+
+	// Run prework.
+	if err := b.processor.PreWork(ctx); err != nil {
+		b.logger.Error("prework failed", "err", err)
+		return
+	}
+
+	// Configure start height.
+	var startHeight uint64
+	latest, err := b.latestProcessedBlock(ctx)
+	switch err {
+	case nil:
+		b.logger.Debug("setting height using latest block", "height", latest+1)
+		startHeight = latest + 1
+	case pgx.ErrNoRows:
+		b.logger.Debug("setting height using range config", "height", b.config.From)
+		startHeight = b.config.From
+	default:
+		b.logger.Error("last block height not found",
+			"err", err,
+		)
+		return
+	}
+
+	// Start a worker to follow and enqueue incoming blocks.
+	go b.followIncoming(ctx, startHeight, b.config.To)
+
+	// Start processing blocks.
+	// XXX: Multiple workers can be started here in future.
+	backoff, err := util.NewBackoff(
+		100*time.Millisecond,
+		6*time.Second, // cap the timeout at the expected consensus block time
+	)
+	if err != nil {
+		b.logger.Error("error configuring indexer backoff policy",
+			"err", err.Error(),
+		)
+		return
+	}
+	for {
+		backoff.Wait()
+
+		b.logger.Info("picking a batch of blocks to process")
+
+		// Pick a batch of blocks to process.
+		heights, err := b.fetchBatchForProcessing(ctx, b.config.From, b.config.To)
+		if err != nil {
+			b.logger.Error("failed to pick blocks for processing",
+				"err", err,
+			)
+			backoff.Failure()
+			continue
+		}
+
+		b.logger.Debug("picked blocks for processing", "heights", heights)
+
+		// Process blocks.
+		for _, height := range heights {
+			b.logger.Info("processing block", "height", height)
+
+			pCtx, cancel := context.WithTimeout(ctx, processBlockTimeout)
+			if err := b.processor.ProcessBlock(pCtx, height); err != nil {
+				cancel()
+				b.logger.Error("error processing block, unlocking",
+					"height", height,
+					"err", err,
+				)
+				// Unlock a failed block, so it can be retired sooner.
+				b.unlockBlock(ctx, height)
+				continue
+			}
+			cancel()
+			b.logger.Info("processed block", "height", height)
+		}
+
+		backoff.Success()
+		b.logger.Info("batch processed")
+
+		// Stop processing if end height is set and was reached.
```

Comment:
A gotcha: If we reach the end(*), it might be the case that some blocks in the from..to range are locked by other analyzers but not yet processed, and they might fail in the future. To ensure that analyzers don't bail too soon, I think we should do another `if (locked and unprocessed blocks exist) { sleep(1min); continue; }` here.

Although maybe it's not needed, what with your mechanism of immediately unlocking failed rows. I'm not sure how it works out. Just something to keep in mind.

(*) Once we have the mega-query, "reaching the end" means the mega-query returns 0 rows, as far as I can tell.

Reply:
The `latestProcessedBlock` query will/should return only processed blocks and not locked unprocessed ones. So I think this is not problematic. 

If the "end block" is currently being processed, the `latestProcessedBlock` will return a lower block, and the loop will not stop here and go into another iteration.

Reply:
>  I think this might be a leftover from your previous approach.

It is. I think I agree with the proposed solution, as it's the most straightforward and a good "sanity-check" that all blocks were indeed processed. 

However, let me reply to some parts of the comment, so that some of the semantics of the "mega-query" are disambiguated.

> For one, does the FirstUnprocessedBlock account for heights that have never been inserted into processed_blocks?

I think that even with the "mega-query", there will be no gaps in inserted `processed_blocks` table. There might be gaps in blocks that were actually processed (there could be failed blocks).

The only way I think this could happen is if there are multiple analyzers running with different from-to ranges. But I don't think even the "mega-query" itself currently handles that case correctly.


> Also, (as a corner case that we're probably OK to not support), suppose we (re)run the analyzer on some range a..b, but previously, analyzers already ran not-to-completion on a range a..c (c>b), and finished some blocks before b and some after b, but with gaps. Then this approach doesn't work

I think that again there would be no gaps in inserted blocks in this case (there would be gaps in actually processed blocks).

> Also, the semantics of latestProcessedBlock() confuse me -- it sometimes returns the latest processed block, and sometimes the earliest UNprocessed block. If we keep this, maybe we can come up with a more meaningful/accurate name for the func?

Yeah, the idea was/is to return the greatest block that has all block up to it indexed. Indeed, the name is poorly chosen, since the initial implementation actually did return the greatest processed block. Should be something in the sense of: `greatestContiguouslyProcessed`.

--- 

I think I might be missing something obvious since both this comment, and the one about deleting rows for failed-to-be-processed blocks, lead me to believe you expect there to be gaps in inserted blocks in the processed blocks table.